### ISSUE TITLE
Extend the "Read active window" command to be able to spell the information, as well as to copy it to clipboard

### ISSUE THREAD

**Issue by zahyur** [nvaccess/nvda#5488]
Sometimes it is nesesary for one to copy the text of the entire window to the clipboard, in order to send it to another person - either as a part of tutorial, or for the other person to be able to help with diagnosing some issue.
Currently, one could enter screen review, go to the top left, mark the position for copying, go to the bottom right and then copy what the screen review sees.
That's not only a lot of keystrokes, but in some situations the review cursor doesn't see some nested elements when copying like that, for example in the CodeFactory's new Eloquence and Vocalizer add-on's Settings dialog.
A possibility of spelling the information also could be useful in some situations, for example when one have to handwrite some information in other place, for example call a phone number, written in some dialog. Even if it's not so widely used, it is good to have it, in order to be consistent with the other commands.
So I propose a double press of the "Read active window" command (NVDA+B) to spell the information and a tripple press to copy it to the clipboard.

**Comment by marrie** [comment]
Why not set a window virtualize keystroke like other screen reading software might have.  Like nvda plus w, or what ever 

Another suggestion would be  nvda plus b, then again to virtualize then you can use your select all key strokes and line selection key strokes to go over and spell the text.

**Comment by josephsl** [comment]
Hi,

 

There is an add-on just for this purpose at:

http://addons.nvda-project.org/addons/virtualRevision.en.html

**Comment by marrie** [comment]
Can this be incorporated in to the nvda source itself so that the user does not have to go grab an add on just to do this? This would help the beginning user of nvda I would think.

**Comment by zahyur** [comment]
Hi,
I know about this add-on, but this ticket is about entirely different thing.
Besides, last time I've checked, that add-on didn't worked under Windows
10 insider priview with NVDA Next.

**Comment by josephsl** [comment]
Hi,

 

I’m afraid not, especially when screen review does not work on modern apps.

**Comment by marrie** [comment]
Ok, in that case I think the ticket is still valid. Why not have a feature where you can go over a window with normal reading keys virtualized like is suggested.

**Comment by zahyur** [comment]
Hi,
I have no problem using neither the review cursor, nor the object navigation.
I don't neither need nor want the screen virtualizing functionality, found in other screen readers.
This ticket has nothing to do with those things, so please refer them to the appropriet tickets.
Look at it like a textual screen shot of the current window, if you wish.
Also many NVDA commands spell the information when doublepress, and copy it to the clipbord on tripplepress. I like this, but unfortunately that's not the case with the "Read current window" command.
There are other commands which don't follow that convention as well, but this ticket is for that one in perticular.
I plan to raise a separate tickets for the others when I have time.

**Comment by zahyur** [comment]
Hi,
I have no problem using neither the review cursor, nor the object navigation.
I don't neither need nor want the screen virtualizing functionality, found in other screen readers.
This ticket has nothing to do with those things, so please refer them to the appropriet tickets.
Look at it like a textual screen shot of the current window, if you wish.
Also many NVDA commands spell the information when doublepress, and copy it to the clipbord on tripplepress. I like this, but unfortunately that's not the case with the "Read current window" command.
There are other commands which don't follow that convention as well, but this ticket is for that one in perticular.
I plan to raise a separate tickets for the others when I have time.
Hope my intentions are more clear this time.

**Comment by marrie** [comment]
Then I per pose a ticket merge as these issues with nvda are the same to me  wrote:
We really have needed a way to virtualize the screen as well as copy that text to clipboard. I can do this in other operating systems and screen readers. We've needed this for years now.

**Comment by derekriemer** [comment]
Do you think using nvda+b thrice would be the behavior you are looking 
for? That would technically be useful.
The problem I see here is that in situations where other screen readers 
can see stuff on screen, nvda can't often. It is because those things 
aren't exposed at all to accessibility api's, so ocr is the only option. 
Can you give me an example app where you would like this behavior?
In many windows dialogs, pressing ctrl+c will copy the output. For 
example, if you go to the nvda about dialog and press ctrl+c it will 
copy this text to the clipboard, which is almost better than a copy from 
the say window command, because this preserves structure. I am not 
saying this wouldn't be a useful feature, but just thought you may want 
to be aware of this already existing feature.

[Window Title]
About NVDA

[Content]
NonVisual Desktop Access (NVDA)
Version: next-12762,4c9504c
URL: http://www.nvaccess.org/
Copyright (C) 2006-2015 NVDA Contributors

NVDA is covered by the GNU General Public License (Version 2). You are 
free to share or change this software in any way you like as long as it 
is accompanied by the license and you make all source code available to 
anyone who wants it. This applies to both original and modified copies 
of this software, plus any derivative works.
For further details, you can view the license from the Help menu.
It can also be viewed online at: 
http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

NVDA is developed by NV Access, a non-profit organisation committed to 
helping and promoting free and open source solutions for blind and 
vision impaired people.
If you find NVDA useful and want it to continue to improve, please 
consider donating to NV Access. You can do this by selecting Donate from 
the NVDA menu.

[OK]

And another dialog (the properties dialog in windows explorer for a folder).
On 11/17/2015 12:17 PM, Zahari Yurukov wrote:
>
> Hi,
> I have no problem using neither the review cursor, nor the object 
> navigation.
> I don't neither need nor want the screen virtualizing functionality, 
> found in other screen readers.
> This ticket has nothing to do with those things, so please refer them 
> to the appropriet tickets.
> Look at it like a textual screen shot of the current window, if you wish.
> Also many NVDA commands spell the information when doublepress, and 
> copy it to the clipbord on tripplepress. I like this, but 
> unfortunately that's not the case with the "Read current window" command.
> There are other commands which don't follow that convention as well, 
> but this ticket is for that one in perticular.
> I plan to raise a separate tickets for the others when I have time.
> Hope my intentions are more clear this time.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5488#issuecomment-157475682>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.drupalgardens.com>
Non-proffessional website. <http://derekriemer.pythonanywhere.com/personal>
Awesome little hand built weather app that rocks! 
<http://derekriemer.pythonanywhere.com/weather>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by bhavyashah** [comment]
Thoughts? @leonardder @Qchristensen @ehollig @dcager

**Comment by leonardder** [comment]
I think both spelling and copying to the clipboard will be a bit hard to implement at first sight, since nvda+b isn't as simple as just reading a piece of text, it uses a generator to read several objects in sequence.

**Comment by derekriemer** [comment]
You could use a compoundDocumentTextInfo or similar

**Comment by Qchristensen** [comment]
The difficulty of implementing aside, I'm not against pressing NVDA+b twice to spell.  As one of the comments note, this would be consistent with other reading commands.

What happens when you press it three times is a tricky one, as with some commands we spell phonetically, and with others, we copy to the clipboard.

Personally, I prefer pressing three times to spell phonetically, as you may accidentally press a command three times which isn't as potentially annoying if it just spells phonetically.

Personally, I wonder if it would be worth having that functionality in general, for pressing a command three times, and having another keystroke to copy the last utterance to the clipboard, whether that was reading a character, or the whole window.

Not sure I helped so much there as added another dimension to the issue and conversation! *smile*

**Comment by bhavyashah** [comment]
I can hardly imagine any usage case for phonetically spelling all contents of the current window, because Caps+B is pretty verbose anyways, and phonetically spelling it might just be excess load for densely populated windows and less responsive speech synthesizers. In contrast, Caps+B to copy all contents of the current window seems far more useful, although the same could also be achieved in many cases by using review modes and NVDA+F9 and F10. Having said that, I remain in favour of copying the contents of the current window (i.e. foreground object) over phonetically spelling it.

**Comment by Qchristensen** [comment]
Good point, I must admit I was thinking overall in terms of consistency of commands when I wrote that, moreso than a specific use for spelling out the entire window.

**Comment by ruifontes** [comment]
Hello!

All that features are not better supplied by Virtual Review addon?

You will get a windows with all text, and you can manage and read it as you 
want/need...

Best regards,

Rui

-----Mensagem Original----- 
De: bhavyashah
Data: 14 de agosto de 2017 03:55
Para: nvaccess/nvda
Cc: Subscribed
Assunto: Re: [nvaccess/nvda] Extend the "Read active window" command to be 
able to spell the information, as well as to copy it to clipboard (#5488)



I can hardly imagine any usage case for phonetically spelling all contents 
of the current window, because Caps+B is pretty verbose anyways, and 
phonetically spelling it might just be excess load for densely populated 
windows and less responsive speech synthesizers. In contrast, Caps+B to copy 
all contents of the current window seems far more useful, although the same 
could also be achieved in many cases by using review modes and NVDA+F9 and 
F10. Having said that, I remain in favour of copying the contents of the 
current window (i.e. foreground object) over phonetically spelling it.

**Comment by Adriani90** [comment]
cc: @paulber007, @michaelDCurran

**Comment by XLTechie** [comment]
I know we're not really doing new features right now, but this request just came up on the nvda@groups.io mailing list again, so wanted to bump this issue.

I do like the consistency of NVDA+b twice to spell (though I can't imagine ever wanting to spell an entire window), and thrice to copy.

However if that is not desirable, perhaps a method could be devised to "copy all speech since last keypress to clipboard", on some assignable command?
